### ISSUE TITLE
Supporting different variant/configuration in metaflow

### ISSUE THREAD

**Issue by abdulsalama** [Netflix/metaflow#430]
We are trying to support use cases of different variants/configuration to metaflow flow. For example, user can have a config file like this declaring some metaflow parameters:

**config1.yaml**
```
Flow_params:
 learning_rate: 0.5
 loss: mae
 export_metrics_keys:
   - r_squared
   - precision
   - recall
```

**config2.yaml**
```
base_yaml: config1.yaml
Flow_params:
 learning_rate: 0.7
```

We want the ability to run the flow with either one of these variant/configuration.

Something like this:
python hello.py --variant config1.yaml

python hello.py --variant config2.yaml

Wondering whats your thoughts on this and what is the best way to support different variants in metaflow

**Comment by abdulsalama** [comment]
Related ticket: https://github.com/Netflix/metaflow/issues/431

**Comment by savingoyal** [comment]
@abdulsalama You can define a parameter that takes in the path to your config file and within your flow you can inspect the contents of your file - 

```
from metaflow import FlowSpec, step, IncludeFile

class VariantFlow(FlowSpec):

    variant = IncludeFile(
        'variant',
        is_text=False,
        required=True)

    @step
    def start(self):
        import json
        self.conf = json.loads(self.variant)
        print(self.conf)
        self.next(self.end)

    @step
    def end(self):
        pass

if __name__ == '__main__':
    VariantFlow()
```

```
python variant_flow.py run --variant variant.json
```

**Comment by abdulsalama** [comment]
thanks @savingoyal. Yeah this would be one way to do it. But two concerns, first this won't work for decorators. Ideally you might have config overrides for decorators and would have them in the same config file. Second, this means losing all the beautiful features of metaflow parameters. For example, how can you now specify defaults for individual parameters? how can you mark some required, and some optional? and if i want to examine the flow using metaflow cli, I cannot get a clean nice list of the parameters.

Also now the semantic is completely different of running the flow with:
python flow.py run --learning-rate=0.5
python flow.py run --variant config.yaml

The flow code will have to be different to support both cases.

Ideally, I was thinking of a more native solution such that if you have a config file like this:

```
resources:
   cpu: 1
   memory: 500 
Flow_params:
 learning_rate: 0.5
 loss: mae

``` 

Then running:
`python flow.py --config-file config.yaml`
is equavilant to
`python flow.py --learning_rate=0.5 --loss=mae with resources:cpu=1,memory=500`

It is more like we expand the config.yaml into the list of parameters.
The flow code won't need to worry about the difference.

**Comment by abdulsalama** [comment]
Also how would we support multiple environments? We can have different config files for each environment. dev.yaml, stage.yaml, and prod.yaml. But that means that for every flow you need 3 config files. Also, the files will need to support some inheritance such that the config is not duplicated.

base.yaml
```
resources:
   cpu: 1
   memory: 500 
Flow_params:
 learning_rate: 0.5
 loss: mae
```

stage.yaml
```
parent_config: base.yaml
Flow_params:
 learning_rate: 0.7

```

Wondering if it make sense for metaflow to support some inheritance resolution (or in more complex cases, Hydra composition)? Also would be great if env variation is part of that config processing/resolution so we can have only one config file instead of # of envs.

**Comment by savingoyal** [comment]
I added [a comment](https://github.com/Netflix/metaflow/issues/431#issuecomment-779352464) in #430 around overriding parameters and decorator attributes in your flows using config files. For managing different configurations, let us know which configuration manager you intend to use, and we can look into providing guidelines for that. We do have a feature in-flight `@project` that will allow users of Metaflow to maintain isolated instances of the same flow - it should be available in the coming weeks.

**Comment by abdulsalama** [comment]
thanks @savingoyal. yes please keep me updated about the @project feature, I'm interested to see how we can utilize this for our use case.

In terms of configuration manager, we are thinking of using Hydra for the very complex cases.

**Comment by savingoyal** [comment]
this is now supported. here is an example - https://outerbounds.com/blog/metaflow-in-notebooks-and-scripts/#running-metaflow-in-scripts-with-custom-configurations

**Comment by savingoyal** [comment]
and more config related goodies - https://docs.metaflow.org/metaflow/configuring-flows/introduction
